File: tools/sync_android_icons.py
"""
安卓原生图标资源同步器（Python 版）

为什么不用 Node 脚本：
  @expo/image-utils 的 borderRadius 依赖 sharp；本机 sharp 不可用，
  会降级到 Jimp，而 Jimp 的 circleAsync 在方形画布上并不能正确裁圆
  （实测 ic_launcher_round.webp 仍是方形）。
  故改用 PIL 精确绘制圆角遮罩，产出可控且与设计稿一致。

产出与 Expo prebuild 的目录/命名/尺寸完全对齐：
  mipmap-{mdpi,hdpi,xhdpi,xxhdpi,xxxhdpi}/
    ic_launcher.webp            legacy 方形，48 * scale
    ic_launcher_round.webp      legacy 圆形，48 * scale
    ic_launcher_foreground.webp adaptive 前景，108 * scale（透明）
    ic_launcher_monochrome.webp adaptive 单色，108 * scale（透明）
  drawable-{...}/splashscreen_logo.png
"""
import os
import sys
from PIL import Image, ImageDraw

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(HERE)
RES = os.path.join(ROOT, "android", "app", "src", "main", "res")
IMAGES = os.path.join(ROOT, "assets", "images")
PREVIEW = os.path.join(HERE, "preview")

BG = (247, 244, 239)          # #F7F4EF，与 app.json / colors.xml 一致
SS = 4                        # 超采样倍数

DPI = [("mdpi", 1), ("hdpi", 1.5), ("xhdpi", 2), ("xxhdpi", 3), ("xxxhdpi", 4)]
LEGACY_BASE = 48
ADAPTIVE_BASE = 108
SPLASH = {"mdpi": 288, "hdpi": 432, "xhdpi": 576, "xxhdpi": 864, "xxxhdpi": 1152}


def rounded(img, size, radius_ratio):
    """把 img 缩放为 size×size，套圆角遮罩，底色 BG。"""
    n = size * SS
    base = Image.new("RGBA", (n, n), BG + (255,))
    fitted = img.resize((n, n), Image.LANCZOS)
    base.alpha_composite(fitted)

    if radius_ratio <= 0:
        return base.resize((size, size), Image.LANCZOS)

    r = n * radius_ratio
    if radius_ratio >= 0.5:
        r = n / 2  # 正圆
    mask = Image.new("L", (n, n), 0)
    ImageDraw.Draw(mask).rounded_rectangle([0, 0, n - 1, n - 1],
                                           radius=int(r), fill=255)
    out = Image.new("RGBA", (n, n), (0, 0, 0, 0))
    out.paste(base, (0, 0), mask)
    return out.resize((size, size), Image.LANCZOS)


def plain(img, size, transparent=True):
    """等比缩放到 size×size；transparent=True 时保留源图透明度，不填底。"""
    n = size * SS
    fitted = img.resize((n, n), Image.LANCZOS)
    if transparent:
        return fitted.resize((size, size), Image.LANCZOS)
    base = Image.new("RGBA", (n, n), BG + (255,))
    base.alpha_composite(fitted)
    return base.resize((size, size), Image.LANCZOS)


def main():
    icon = Image.open(os.path.join(IMAGES, "icon.png")).convert("RGBA")
    fg = Image.open(os.path.join(IMAGES, "android-icon-foreground.png")).convert("RGBA")
    mono = Image.open(os.path.join(IMAGES, "android-icon-monochrome.png")).convert("RGBA")
    splash = Image.open(os.path.join(IMAGES, "splash-icon.png")).convert("RGBA")

    print(f"底色 #F7F4EF  超采样 {SS}x")
    print()

    for key, scale in DPI:
        d = os.path.join(RES, f"mipmap-{key}")
        os.makedirs(d, exist_ok=True)
        legacy = round(LEGACY_BASE * scale)
        adapt = round(ADAPTIVE_BASE * scale)

        # legacy 方形（微圆角 22.46%，与设计稿一致；启动器会再裁）
        rounded(icon, legacy, 0.2246).save(os.path.join(d, "ic_launcher.webp"))
        # legacy 圆形
        rounded(icon, legacy, 0.5).save(os.path.join(d, "ic_launcher_round.webp"))
        # 自适应前景 / 单色（透明）
        plain(fg, adapt, transparent=True).save(
            os.path.join(d, "ic_launcher_foreground.webp"))
        plain(mono, adapt, transparent=True).save(
            os.path.join(d, "ic_launcher_monochrome.webp"))

        print(f"  mipmap-{key:9s} launcher {legacy:>3}px  圆 {legacy:>3}px  "
              f"前景 {adapt:>3}px  单色 {adapt:>3}px")

    print()
    for key, size in SPLASH.items():
        d = os.path.join(RES, f"drawable-{key}")
        os.makedirs(d, exist_ok=True)
        plain(splash, size, transparent=True).save(
            os.path.join(d, "splashscreen_logo.png"))
        print(f"  drawable-{key:8s} splashscreen_logo {size:>4}px")

    # 校验图：圆形 + 方形 + 前景
    os.makedirs(PREVIEW, exist_ok=True)
    d196 = os.path.join(RES, "mipmap-xhdpi")
    tiles = [
        Image.open(os.path.join(d196, "ic_launcher.webp")).convert("RGBA"),
        Image.open(os.path.join(d196, "ic_launcher_round.webp")).convert("RGBA"),
        Image.open(os.path.join(d196, "ic_launcher_foreground.webp")).convert("RGBA"),
    ]
    K = 3
    W = sum(t.width * K + 24 for t in tiles) + 24
    H = max(t.height for t in tiles) * K + 48
    c = Image.new("RGB", (W, H), (36, 30, 25))
    x = 24
    for t in tiles:
        b = t.resize((t.width * K, t.height * K), Image.NEAREST)
        c.paste(b, (x, 24), b)
        x += b.width + 24
    c.save(os.path.join(PREVIEW, "native_check.png"))
    print(f"\n校验图 -> {os.path.join(PREVIEW, 'native_check.png')}")
    print("完成。")


if __name__ == "__main__":
    main()

File: tools/_archive/sync_android_icons.js
/**
 * 安卓原生图标资源同步器
 *
 * 为什么不直接跑 `expo prebuild`：
 *   用户正在改别的功能，此时 prebuild --clean 会重建整个 android 目录，
 *   可能冲掉手工改动。本脚本只重写图标相关的资源文件，其余一律不动。
 *
 * 算法来源：@expo/prebuild-config/build/plugins/icons/withAndroidIcons.js
 *   - legacy 基准 48px / adaptive 基准 108px
 *   - 密度倍率 mdpi 1 / hdpi 1.5 / xhdpi 2 / xxhdpi 3 / xxxhdpi 4
 *   - ic_launcher        : 主图标 + iconBackground 底色，resizeMode cover
 *   - ic_launcher_round  : 同图，borderRadius = size * 0.5
 *   - ic_launcher_foreground / monochrome : 透明底，按 adaptive 基准
 *   - splashscreen_logo  : 启动图，drawable-* 各档
 * 本脚本复用 @expo/image-utils 的 generateImageAsync，产出与 prebuild 一致。
 */
const fs = require('fs');
const path = require('path');

const { generateImageAsync } = require('@expo/image-utils');

const ROOT = path.resolve(__dirname, '..');
const RES = path.join(ROOT, 'android', 'app', 'src', 'main', 'res');
const IMAGES = path.join(ROOT, 'assets', 'images');

// 与 app.json 保持一致
const APP = require(path.join(ROOT, 'app.json')).expo;
const BG_COLOR = APP.android.adaptiveIcon.backgroundColor; // #F7F4EF

const DPI = [
  { key: 'mdpi', scale: 1 },
  { key: 'hdpi', scale: 1.5 },
  { key: 'xhdpi', scale: 2 },
  { key: 'xxhdpi', scale: 3 },
  { key: 'xxxhdpi', scale: 4 },
];

const LEGACY_BASE = 48;    // ic_launcher / ic_launcher_round
const ADAPTIVE_BASE = 108; // foreground / monochrome

// 启动图基准（Expo splash 插件用 160dp 档位近似，这里沿用现有文件的尺寸关系）
const SPLASH_DPI = {
  mdpi: 288,
  hdpi: 432,
  xhdpi: 576,
  xxhdpi: 864,
  xxxhdpi: 1152,
};

async function gen(src, size, { backgroundColor = 'transparent', borderRadius } = {}) {
  const res = await generateImageAsync(
    { projectRoot: ROOT, cacheType: 'gewu-icon-sync' },
    {
      src,
      width: size,
      height: size,
      resizeMode: 'cover',
      backgroundColor,
      borderRadius,
    }
  );
  return res.source;
}

async function main() {
  const icon = path.join(IMAGES, 'icon.png');
  const fg = path.join(IMAGES, 'android-icon-foreground.png');
  const mono = path.join(IMAGES, 'android-icon-monochrome.png');
  const splash = path.join(IMAGES, 'splash-icon.png');

  for (const f of [icon, fg, mono, splash]) {
    if (!fs.existsSync(f)) throw new Error('缺素材: ' + f);
  }

  console.log('底色 iconBackground ->', BG_COLOR);
  console.log('');

  for (const { key, scale } of DPI) {
    const dir = path.join(RES, `mipmap-${key}`);
    fs.mkdirSync(dir, { recursive: true });

    // 1. legacy 方形图标（圆角由启动器裁切，这里铺满方形）
    const legacySize = Math.round(LEGACY_BASE * scale);
    const legacy = await gen(icon, legacySize, { backgroundColor: BG_COLOR });
    fs.writeFileSync(path.join(dir, 'ic_launcher.webp'), legacy);

    // 2. legacy 圆形图标
    const round = await gen(icon, legacySize, {
      backgroundColor: BG_COLOR,
      borderRadius: legacySize * 0.5,
    });
    fs.writeFileSync(path.join(dir, 'ic_launcher_round.webp'), round);

    // 3. 自适应前景（透明）
    const adSize = Math.round(ADAPTIVE_BASE * scale);
    const advFg = await gen(fg, adSize, { backgroundColor: 'transparent' });
    fs.writeFileSync(path.join(dir, 'ic_launcher_foreground.webp'), advFg);

    // 4. 主题单色图标（透明）
    const advMono = await gen(mono, adSize, { backgroundColor: 'transparent' });
    fs.writeFileSync(path.join(dir, 'ic_launcher_monochrome.webp'), advMono);

    console.log(
      `  mipmap-${key.padEnd(8)} ic_launcher ${legacySize}px / round ${legacySize}px / fg+mono ${adSize}px`
    );
  }

  // 5. 启动图
  for (const [key, size] of Object.entries(SPLASH_DPI)) {
    const dir = path.join(RES, `drawable-${key}`);
    fs.mkdirSync(dir, { recursive: true });
    const logo = await gen(splash, size, { backgroundColor: 'transparent' });
    fs.writeFileSync(path.join(dir, 'splashscreen_logo.png'), logo);
    console.log(`  drawable-${key.padEnd(8)} splashscreen_logo ${size}px`);
  }

  console.log('\n完成。原生资源已同步为 v3 图标。');
}

main().catch((e) => {
  console.error('失败:', e.message);
  process.exit(1);
});
